api/src/main/java/net/kyori/adventure/bossbar/AbstractBossBar.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * This file is part of adventure, licensed under the MIT License.
+ *
+ * Copyright (c) 2017-2020 KyoriPowered
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.bossbar;
+
+import java.util.stream.Stream;
+import net.kyori.adventure.text.Component;
+import net.kyori.examination.Examinable;
+import net.kyori.examination.ExaminableProperty;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractBossBar implements BossBar, Examinable {
+  private Component name;
+  private float percent;
+  private Color color;
+  private Overlay overlay;
+  private boolean darkenScreen;
+  private boolean playBossMusic;
+  private boolean createWorldFog;
+
+  protected AbstractBossBar(final @NonNull Component name, final float percent, final @NonNull Color color, final @NonNull Overlay overlay) {
+    this.name = requireNonNull(name, "name");
+    this.percent = percent;
+    this.color = requireNonNull(color, "color");
+    this.overlay = requireNonNull(overlay, "overlay");
+  }
+
+  @Override
+  public @NonNull Component name() {
+    return this.name;
+  }
+
+  @Override
+  public @NonNull BossBar name(final @NonNull Component name) {
+    if(name != this.name) {
+      this.name = requireNonNull(name, "name");
+      this.changed(Change.NAME);
+    }
+    return this;
+  }
+
+  @Override
+  public float percent() {
+    return this.percent;
+  }
+
+  @Override
+  public @NonNull BossBar percent(final float percent) {
+    if(percent != this.percent) {
```

Comment:
An optimization hint here: often boss bars are used for long countdowns (say 30 minutes or 2 hours). If `percent` changes from `0.0100` to `0.01001` there is no *effective* difference any client could ever see. Yet, an updated packet would be sent anyway.

To reduce packet verbosity: I would recommend defining an `static final int PERCENT_EPSILON = 0.01`, if the change in percent is less than that, the class updates the field, but `changed` is not called.

I would put this in the abstract implementation, because I believe this would benefit and be useful by all implementations.